### PR TITLE
Conv nonlinearity cost

### LINKED ISSUE #1088: Incorporate cost in ConvNonLinearity
The cost should be a default suitable for that non-linearity, and `ConvElemwise.cost` should call the appropriate one, instead of working only on sigmoid.
See https://github.com/lisa-lab/pylearn2/issues/1087

### PR DESCRIPTION
**Pull request by lamblin**
Rebase of #1513 and #1547 by @matrogers.
Fixes #1088.

### CODE REVIEW

File: pylearn2/models/mlp.py [hunk 1]
```diff
@@ -2734,6 +2764,11 @@ def apply(self, linear_response):
             linear_response * (linear_response < 0.)
         return p
 
+    @wraps(ConvNonlinearity.cost)
+    def cost(self, Y, Y_hat, batch_axis):
+        raise NotImplementedError(
+            str(type(self)) + " does not implement cost function.")
```
Comment (bouthilx) [inline pylearn2/models/mlp.py]: Isn't the same as ConvNonlinearity.cost? Why overwriting it then?

Comment (lamblin) [inline pylearn2/models/mlp.py]: Right, we can remove that.

File: pylearn2/models/tests/test_convelemwise_cost.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                                num_channels=1),
+        layers=[ConvElemwise(layer_name='conv',
+                             nonlinearity=conv_nonlinearity,
+                             output_channels=1,
+                             kernel_shape=[1, 1],
+                             pool_shape=[1, 1],
+                             pool_stride=[1, 1],
+                             irange=1.0)],
+        batch_size=1
+    )
```
Comment (lamblin) [inline pylearn2/models/tests/test_convelemwise_cost.py]: This currently fails, we should take code from the other `check_` method instead.
Also, we should check that the output is the same, even if the cost is not implemented.

**Comment by lamblin** [comment]
The new tests pass, and I fixed an issue with conda stalling when trying to solve the dependencies for package installation.
The remaining issues are due to the fact that pylearn2 depends on an un-maintained copy of `scan` that was in `theano.sandbox`, and that has been removed from the development version.

**Comment by nouiz** [comment]
This PR changed groundhog to use the current scan by just changing the import:

https://github.com/lisa-groundhog/GroundHog/pull/35

This can probably be done here too.

**Comment by lamblin** [comment]
Travis is passing now.

**Pull request merged**